### ISSUE TITLE
fully_connected (per-channel, ESP32-S3): incorrect memory-layout-dependent results for narrow output width (out_ch = 2)

### ISSUE THREAD

**Issue by davidallenmann** [espressif/esp-nn#34]
### Checklist

- [x] Checked the issue tracker for similar issues to ensure this is not a duplicate
- [x] Read the documentation to confirm the issue is not addressed there and your configuration is set correctly
- [x] Tested with the latest version to ensure the issue hasn't been fixed

### How often does this bug occurs?

always

### Expected behavior

FULLY_CONNECTED produces the same results as the TFLM reference kernel and the
TFLite Python interpreter for all layer geometries — or, if the esp-nn
optimized kernel cannot handle a geometry, the kernel wrapper falls back to
the reference implementation instead of producing silently wrong results.

esp-nn documents the constraint explicitly (include/esp_nn_esp32s3.h, above
both `esp_nn_fully_connected_s8_esp32s3` and
`esp_nn_fully_connected_per_ch_s8_esp32s3`):

> Current version works only on aligned input.
> row_len and channels should both be multiple of 8.

But `tensorflow/lite/micro/kernels/esp_nn/fully_connected.cc` calls these
kernels unconditionally — there is no multiple-of-8 check and no fallback.
Any int8 model whose FC output width is not a multiple of 8 (binary
classifiers with 2 outputs especially) gets silently corrupted logits.

For our layer — int8 FULLY_CONNECTED, input 1x1280, weights 2x1280
per-channel quantized (2 scales), followed by SOFTMAX(1,2) — the TFLite
Python interpreter on the identical .tflite and identical input produces raw
int8 softmax outputs `[117, -117]` (class 0 at ~96% confidence). The device
should match this.

## Actual behavior (suspected bug)

On ESP32-S3 (octal PSRAM; model buffer and tensor arena in PSRAM; ESP-IDF
v5.5.3; `CONFIG_NN_OPTIMIZED=y`; esp-nn 1.1.0 and 1.1.1 from the ESP
Component Registry, git tags v1.1.0/v1.1.1, also reproduced on 1.2.5; called
via esp-tflite-micro 1.3.2 and a 1.3.3~1-era snapshot), the same model and
same input produce **incorrect logits that depend on memory layout**:

| configuration | raw int8 softmax output [class0, class1] |
|---|---|
| TFLite Python reference (host) | `[117, -117]` |
| S3, all esp-nn ops, build A | `[0, 0]` (exact tie, softmax 0.500/0.500) |
| S3, all esp-nn ops, build B (unrelated code changes) | `[126, -126]` |
| S3, all esp-nn ops, build C (model moved PSRAM → flash rodata) | `[-126, 126]` (confidently reversed) |
| S3, `CONFIG_NN_ANSI_C=y` (all reference kernels) | correct, matches host |
| S3, **only** `kernels/esp_nn/fully_connected.cc` excluded, all other esp-nn ops active | `[118, -118]` — correct |

Evidence that isolates the bug to the per-channel FC kernel:

- The input tensor is byte-verified: on-device min/max/mean of the 32,768-byte
  int8 input matches the host pipeline exactly for every test case.
- The model bytes are byte-verified: CRC32 of the model buffer computed
  on-device (0x11CF319D, 3,953,856 bytes) matches the host file.
- Results are deterministic within a build: invoking 3× on the same restored
  input gives byte-identical outputs — not a race or timing issue. But any
  unrelated change that shifts buffer addresses changes the (wrong) result,
  which suggests an out-of-bounds read near the 2-element output.
- Per-op bisection: with esp-nn `add`, `mul`, `softmax` forced to reference
  kernels and FC still optimized → still wrong. Adding `fully_connected` to
  the reference set → correct. Restoring everything except FC to esp-nn →
  still correct. FULLY_CONNECTED is the only defective op for this model.
- Buffer alignment is not the cause: reproduced with model and arena both
  16-byte and 64-byte aligned (`heap_caps_aligned_alloc`).
- Our other products run models with **per-tensor** quantized FC heads
  (96 outputs) on the same chip and config with no issue — they never enter
  the per-channel path. Suspect file:
  `src/fully_connected/esp_nn_fully_connected_per_ch_s8_esp32s3.S`
  (added in 966982f). Compare db4b054 ("fully_connected: Fix mismatch for
  input sizes < 8") — this looks like the same class of small-dimension edge
  case on the newer per-channel path.

Suggested minimal repro for the esp-nn test_app:
`esp_nn_fully_connected_per_ch_s8` with in_features=1280, out_ch=2
(2 per-channel scales) compared against the ANSI implementation over
randomized inputs and varying buffer placement/padding.

Workaround: excluding `tensorflow/lite/micro/kernels/esp_nn/fully_connected.cc`
from the esp-tflite-micro build (FC falls back to the TFLM reference kernel,
negligible cost at 2×1280 MACs) fully restores correctness while keeping all
other esp-nn optimizations.

Impact: Model is an
EfficientNet-Lite0-style int8 audio classifier
(`... → MEAN → FULLY_CONNECTED 1280→2 → SOFTMAX`); we can share the .tflite
and a captured input tensor privately on request.

### Error logs or terminal output

```shell
Our test harness feeds a fixed, host-verified input tensor and invokes 3x on
the same input (`out=` shows raw int8 outputs of each of the 3 invokes;
`in[..]` is input min/max/mean, identical to host in all runs).

All esp-nn ops enabled — confidently reversed on a 96%-confidence example,
exact tie on another:


I (31715) ML: test rices_whale/8CBFEA0B65BC_00000_251224T010041.wav w0: 10840490_rices_whale=0.008 SYS_background=0.992 in[-3..127 mean=3.0 zp=3] out=[-126,126] [-126,126] [-126,126] (3703 ms) FAIL
I (75475) ML: test SYS_background/8CBFEA0B65BC_00001_251224T045742.wav w0: 10840490_rices_whale=0.500 SYS_background=0.500 in[-88..61 mean=3.0 zp=3] out=[0,0] [0,0] [0,0] (2593 ms) TIE


CONFIG_NN_ANSI_C=y — correct, matches host:


I (66953) ML: test rices_whale/8CBFEA0B65BC_00000_251224T010041.wav w0: 10840490_rices_whale=0.992 SYS_background=0.008 in[-3..127 mean=3.0 zp=3] out=[126,-126] [126,-126] [126,-126] (14739 ms) PASS
I (125353) ML: test rices_whale/8CBFEA0B65BC_00008_251224T104738.wav w0: 10840490_rices_whale=0.992 SYS_background=0.008 in[-64..90 mean=3.0 zp=3] out=[126,-126] [126,-126] [126,-126] (14739 ms) PASS
I (183753) ML: test SYS_background/8CBFEA0B65BC_00001_251224T045742.wav w0: 10840490_rices_whale=0.008 SYS_background=0.992 in[-88..61 mean=3.0 zp=3] out=[-126,126] [-126,126] [-126,126] (14739 ms) PASS


Only `kernels/esp_nn/fully_connected.cc` excluded, every other esp-nn op
optimized — correct, matches host, near-full esp-nn speed:


I (31395) ML: test rices_whale/8CBFEA0B65BC_00000_251224T010041.wav w0: 10840490_rices_whale=0.906 SYS_background=0.094 in[-3..127 mean=3.0 zp=3] out=[104,-104] [104,-104] [104,-104] (3696 ms) PASS
I (56655) ML: test rices_whale/8CBFEA0B65BC_00008_251224T104738.wav w0: 10840490_rices_whale=0.961 SYS_background=0.039 in[-64..90 mean=3.0 zp=3] out=[118,-118] [118,-118] [118,-118] (3696 ms) PASS
I (81925) ML: test SYS_background/8CBFEA0B65BC_00001_251224T045742.wav w0: 10840490_rices_whale=0.023 SYS_background=0.977 in[-88..61 mean=3.0 zp=3] out=[-122,122] [-122,122] [-122,122] (3696 ms) PASS
I (81935) ML: test summary: 10840490_rices_whale 2/2 (0 tie)
I (81935) ML: test summary: SYS_background 1/1 (0 tie)
I (81945) ML: test summary: TOTAL 3/3 (0 tie)


Host TFLite Python reference for the same three files (raw output bytes):
`[93, -93]`, `[117, -117]`, `[-119, 119]` — matching both correct device
configurations within quantization rounding.
```

### Steps to reproduce the behavior

1. Take any int8 post-training-quantized classifier whose final
   FULLY_CONNECTED has 2 output channels (a binary head, e.g. 1280->2). The
   standard TFLite converter produces per-channel FC weights by default; the
   bug reproduces with per-tensor FC as well (both S3 kernels carry the same
   constraint).
2. Build for ESP32-S3 with esp-tflite-micro + esp-nn, `CONFIG_NN_OPTIMIZED=y`.
3. Feed a fixed input tensor and record the raw int8 output of the final
   softmax (or the FC logits directly).
4. Run the identical .tflite on the identical input tensor with the TFLite
   Python interpreter on a host.
5. Compare: device output differs from host (exact-tie `[0,0]` or
   sign-reversed logits, depending on memory layout).
6. Rebuild with `CONFIG_NN_ANSI_C=y`, or delete
   `tensorflow/lite/micro/kernels/esp_nn/fully_connected.cc` so FC uses the
   TFLM reference kernel: device now matches the host exactly.
7. Optional, kernel-level: run the standalone program below, which compares
   the optimized S3 FC kernels against esp-nn's own `_ansi` implementations
   at out_channels=2 and reports mismatches.

### Standalone reproduction program

Minimal ESP-IDF app_main for an ESP32-S3 project with the esp-nn component
(`CONFIG_NN_OPTIMIZED=y`). It compares `esp_nn_fully_connected_s8` and
`esp_nn_fully_connected_per_ch_s8` (mapped to the S3 assembly) against the
`_ansi` reference implementations shipped in the same library, on random
inputs with out_channels=2, and also checks guard bytes around the 2-byte
output for out-of-bounds writes. (Compiles warning-free against the esp-nn
1.1.1 headers; the `_ansi` sources are in esp-nn's unconditional source list
so they link in optimized builds, and `esp_nn_esp32s3.h` maps the plain
names to the S3 assembly. Not executed on hardware on our side — our
verified evidence is the system-level reproduction above.)

```c
#include <stdio.h>
#include <string.h>
#include <stdint.h>
#include "esp_nn.h"
#include "esp_nn_ansi_headers.h"   // _ansi reference prototypes

#define ROW_LEN 1280               // multiple of 8 (constraint satisfied)
#define OUT_CH  2                  // NOT a multiple of 8 (constraint violated)
#define GUARD   32
#define TRIALS  200

static int8_t  input[ROW_LEN] __attribute__((aligned(16)));
static int8_t  filter[OUT_CH * ROW_LEN] __attribute__((aligned(16)));
static int32_t bias[OUT_CH];
static int8_t  out_region[GUARD + OUT_CH + GUARD];  // guarded output
static int8_t  out_ref[OUT_CH];

static uint32_t rng = 12345;
static int8_t rnd8(void) { rng = rng * 1664525u + 1013904223u; return (int8_t)(rng >> 24); }

void app_main(void)
{
    int pt_mismatch = 0, pc_mismatch = 0, oob = 0;

    for (int t = 0; t < TRIALS; t++) {
        for (int i = 0; i < ROW_LEN; i++)          input[i]  = rnd8();
        for (int i = 0; i < OUT_CH * ROW_LEN; i++) filter[i] = rnd8();
        for (int c = 0; c < OUT_CH; c++)           bias[c]   = (int32_t)rnd8() * 64;

        const int32_t input_offset  = -3;     // values from our real model
        const int32_t filter_offset = 0;
        const int32_t out_offset    = -128;
        const int32_t mult          = 1355715584;   // representative multiplier
        const int32_t shift         = -6;
        const int32_t mults[OUT_CH]  = { mult, mult - 12345 };
        const int32_t shifts[OUT_CH] = { shift, shift };

        // ---- per-tensor kernel ----
        memset(out_region, 0xA5, sizeof out_region);
        int8_t *out = &out_region[GUARD];
        esp_nn_fully_connected_s8(input, input_offset, ROW_LEN, filter,
                                  filter_offset, bias, out, OUT_CH,
                                  out_offset, shift, mult, -128, 127);
        esp_nn_fully_connected_s8_ansi(input, input_offset, ROW_LEN, filter,
                                       filter_offset, bias, out_ref, OUT_CH,
                                       out_offset, shift, mult, -128, 127);
        if (memcmp(out, out_ref, OUT_CH) != 0) {
            if (++pt_mismatch <= 5)
                printf("per-tensor trial %d: opt [%d,%d]  ansi [%d,%d]\n",
                       t, out[0], out[1], out_ref[0], out_ref[1]);
        }

        // ---- per-channel kernel ----
        memset(out_region, 0xA5, sizeof out_region);
        esp_nn_fully_connected_per_ch_s8(input, input_offset, ROW_LEN, filter,
                                         filter_offset, bias, out, OUT_CH,
                                         out_offset, shifts, mults, -128, 127);
        esp_nn_fully_connected_per_ch_s8_ansi(input, input_offset, ROW_LEN,
                                              filter, filter_offset, bias,
                                              out_ref, OUT_CH, out_offset,
                                              shifts, mults, -128, 127);
        if (memcmp(out, out_ref, OUT_CH) != 0) {
            if (++pc_mismatch <= 5)
                printf("per-channel trial %d: opt [%d,%d]  ansi [%d,%d]\n",
                       t, out[0], out[1], out_ref[0], out_ref[1]);
        }

        for (int i = 0; i < GUARD; i++) {
            if (out_region[i] != (int8_t)0xA5 ||
                out_region[GUARD + OUT_CH + i] != (int8_t)0xA5) { oob++; break; }
        }
    }

    printf("row_len=%d out_ch=%d: per-tensor mismatches %d/%d, "
           "per-channel mismatches %d/%d, guard corruption in %d trials\n",
           ROW_LEN, OUT_CH, pt_mismatch, TRIALS, pc_mismatch, TRIALS, oob);
}
```

### Project release version

v1.1.1

### System architecture

Intel/AMD 64-bit (modern PC, older Mac)

### Operating system

MacOS

### Operating system version

Sequoia 15.6

### Shell

ZSH

**Comment by vikramdattu** [comment]
@davidallenmann  thanks for the thorough report!

We reproduced a real bug in the ESP32-S3 FULLY_CONNECTED kernels and have a fix up. It will ship in esp-nn 1.2.6. The root cause is different from the one you inferred, in a way that matters for your model.

It is not out_ch granularity. The row_len and channels should both be multiple of 8 note in esp_nn_esp32s3.h is stale (It stopped describing the code when the multi-path C dispatcher landed in 1.2.4), which loops one output channel at a time and has a scalar remainder loop for row_len. We swept out_ch ∈ {1, 2, 3, 7, 8, 15, 16} at row_len 1280 on hardware: with a 16-byte aligned input, all pass, including 2. With the input shifted 4 bytes, all fail, including 8 and 16. So a multiple-of-8 guard in fully_connected.cc would not have fixed this, and would have de-optimized many working models. We've removed the misleading note.

The trigger is input alignment. Sweeping input misalignment 0–15 (row_len 1280, out_ch 2, 8 randomized trials each, mismatches vs the _ansi reference)!

Two of your hypotheses don't survive: I saw no out-of-bounds write (guard bytes around the output were intact in every configuration), and the wrong values saturate both channels to the same value, which is a corrupted accumulator rather than an OOB read near a narrow output.

On the attached reproduction program: as written it declares `static int8_t input[ROW_LEN] __attribute__((aligned(16)));`, which forces the fast path. Run on an S3 it gives 0/200 mismatches on both kernels. Changing only the input pointer to input + 4 in the same harness gives 155/200. That's consistent with your note that it hadn't been executed on hardware, and it's worth knowing if you reuse it.

TFLite Micro aligns arena buffers to 16 bytes () in micro_arena_constants.h), so an FC input arriving from a normal arena tensor should already take the correct fast path. That means, we have fixed a genuine issue, but not as per your report. Request you to confirm if your model inference passes with 1.2.6 release, if it still bites you, do share the reproducible test case.

Your workaround (excluding fully_connected.cc so FC falls back to the TFLM reference costs almost nothing at 2×1280 MACs, so it's fine to stay on it in the meantime.

**Comment by davidallenmann** [comment]
And Claude says :)
Thanks for the fix and for the correction on the repro — you're right that the `aligned(16)` input forced the fast path; good catch, and the +4 result is convincing for the 1.2.4+ dispatcher.

One clarification that may matter for whether 1.2.6 closes our case: all of our failing configurations ran esp-nn **1.1.0 and 1.1.1**, i.e. the pre-dispatcher `.S` kernels (`esp_nn_fully_connected_s8_esp32s3.S` and `esp_nn_fully_connected_per_ch_s8_esp32s3.S`), not the multi-path C dispatcher your sweep exercised. Per your note, TFLM arena tensors are 16-byte aligned — so our FC input (the output of MEAN, an arena tensor) should have been aligned in every failing run, yet out_ch=2 was consistently wrong and became correct with only `kernels/esp_nn/fully_connected.cc` excluded. So on 1.1.x there appears to be a distinct defect from the one you fixed.

One possibly relevant variable on our side: the *filter* pointer's alignment did vary across our configurations (16-byte-aligned PSRAM copy vs an `alignas(8)` flash array), and the two produced different wrong outputs — if the 1.1.x assembly (or the current dispatcher) assumes filter alignment as well as input alignment, that would fit our data.

We'll validate 1.2.6 against our on-device regression suite (same model, fixed host-verified inputs, byte-compared against the TFLite Python reference) as soon as it's on the registry and report back either way. If it still reproduces, we'll send the .tflite plus a captured input tensor and the failing/expected logits.

**Comment by vikramdattu** [comment]
@davidallenmann request you to try your model on the latest 1.2.6, I am hoping this will now fix your issue. If it still persist, I will need to reproduce the bug and take that up as a follow-up fix. (with your shared sample, the issue did not reproduce).

You may also align the filter buffers in place to have a better performance(esp-nn won't need to do the alignment juggling in that case). You need to run the script provided [here](https://github.com/tensorflow/tflite-micro/blob/main/tensorflow/lite/micro/tools/tflite_flatbuffer_align.py). Hope that helps
